### PR TITLE
python/grains: add mentoring notes

### PR DESCRIPTION
**Pull request by Tarun193**
Added mentoring notes for the python grain exercise

### CODE REVIEW

File: tracks/python/exercises/grains/mentoring.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+## Reasonable solution By calculating power.
+
+```
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:3]: Give the language for the fence blocks…

~~~suggestion
```python
~~~

File: tracks/python/exercises/grains/mentoring.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+as we did 2**(number-1) on block one there will be only 1 grain, 2 power 0 = 1.
+on block 2 there will 2 and on block 3 there will be 4 which is equal to 2 power 1 and 2 power 2 respectively.
+Hence gerenalized **2 power (number-1)**.
+
+### For function total():
+there is a gerenal formula:
+**2^0+2^1+2^2+2^3+....+2^n=(2^(n+1))-1**.
+
+## Solution using Bitwise left operator
+```
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:35]:
```suggestion
```python
```

File: tracks/python/exercises/grains/mentoring.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        return 2<<(number-1)
+
+
+def total():
+    #as 2^0+2^1+2^2+2^3+....2^n=(2^(n+1))-1
+    return 2<<(64)-1
+```
+Left shift bitwise operator
+
+1<<0 --> int binary 0 0 1 --> decimal 1 =2^0
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:49]: Perhaps give some space around the operators, for readability.

Comment (kotp) [reply tracks/python/exercises/grains/mentoring.md:49]: This section might be more presentable as an example using REPL syntax, and code block fencing?

**Review by kotp — CHANGES_REQUESTED** [review]

File: tracks/python/exercises/grains/mentoring.md (hunk 3 above)
Comment (Tarun193) [reply tracks/python/exercises/grains/mentoring.md]: Now, check if it is okey?

File: tracks/python/exercises/grains/mentoring.md [hunk 4]
```diff
@@ -44,12 +44,20 @@ def total():
...
+### Left shift bitwise operator
 
-1<<0 --> int binary 0 0 1 --> decimal 1 =2^0
-
-1<<1 --> int binary 0 1 0 --> decimal 2 =2^1
-
-1<<2 --> int binary 1 0 0 --> decimal 4 =2^2
-
-1<<n --> =2^n
+**2 power n = 1<<n**
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:49]: Space around the operators make for easier reading.

```suggestion
**2 power n = 1 << n** 
```

File: tracks/python/exercises/grains/mentoring.md [hunk 5]
```diff
@@ -44,12 +44,20 @@ def total():
...
+>>> 1<<0  # 1<<1  in binary -> 1
+1
+>>> 2^1
+2
+>>> 1<<1 # 1<<1  in binary -> 1 0
+2
+>>> 2^2
+4
+>>> 1<<2 # 1<<1  in binary -> 1 0 0
+4
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:62]:
```suggestion
>>> 2^0
1
>>> 1 << 0  # 1 <<1   in binary -> 1
1
>>> 2^1
2
>>> 1 << 1 # 1 << 1  in binary -> 1 0
2
>>> 2^2
4
>>> 1 << 2 # 1 << 1  in binary -> 1 0 0
4
```

**Review by kotp — CHANGES_REQUESTED** [review]

File: tracks/python/exercises/grains/mentoring.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
+## Reasonable solution By calculating power.
+
+```python
+def square(number):
+    """
+    This Function is going to return the number of grains on a given square number.
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:6]:
```suggestion
    This Function must return the number of grains on a given square number.
```

File: tracks/python/exercises/grains/mentoring.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    """
+    if(number<1 or number>64):
+        raise ValueError("square must be between 1 and 64")
+    else:
+        return 2**(number-1)
+
+
+def total():
+    """
+    This Function is going to return the total number of grains on the chess board.
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:18]:
```suggestion
    This Function must return the total number of grains on the chess board.
```

File: tracks/python/exercises/grains/mentoring.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    :return int total number of grains a chess board can contain.
+    """
+    #as 2^0+2^1+2^2+2^3+....2^n=(2^(n+1))-1
+    return 2**(64)-1
+```
+
+### For function square(number):
+as we did 2**(number-1) on block one there will be only 1 grain, 2 power 0 = 1.
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:26]:
```suggestion
as we did `2**(number -`1)` on block one there will be only 1 grain, 2 to the power of 0 equals 1.
```

File: tracks/python/exercises/grains/mentoring.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+on block 2 there will 2 and on block 3 there will be 4 which is equal to 2 power 1 and 2 power 2 respectively.
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:27]:
```suggestion
on block 2 there will 2 and on block 3 there will be 4 which is equal to 2 to the power of 1, and 2 to the power of 2, respectively.
```

File: tracks/python/exercises/grains/mentoring.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+Hence gerenalized **2 power (number-1)**.
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:28]:
```suggestion
Hence generalized, **2 power (number-1)**.
```

File: tracks/python/exercises/grains/mentoring.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+### For function total():
+there is a gerenal formula:
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:31]:
```suggestion
there is a general formula:
```

But is it a general formula, or is it sufficient to just state that we have the formula, without the "general" word being used?

File: tracks/python/exercises/grains/mentoring.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+**2^0+2^1+2^2+2^3+....+2^n=(2^(n+1))-1**.
+
+## Solution using Bitwise left operator
+```python
+def square(number):
+    if(number<1 or number>64):
+        raise ValueError("square must be between 1 and 64")
+    else:
+        return 2<<(number-1)
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:40]: Is it customary to squeeze the operations together without space?  I find it cluttered, and harder to read (as my eyes have a hard time differentiating from each "word" without the space).

**Review by kotp — CHANGES_REQUESTED** [review]

File: tracks/python/exercises/grains/mentoring.md [hunk 13]
```diff
@@ -3,7 +3,7 @@
 ```python
 def square(number):
     """
-    This Function is going to return the number of grains on a given square number.
+    This Function must return the number of grains on a given square number.
     :parama: int square number on which we want to calculate the amount of grains.
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:7]:
```suggestion
    :param: int square number on which we want to calculate the amount of grains.
```

**Review by kotp — CHANGES_REQUESTED** [review]
The last thing, I think, from me for this.

**Comment by Tarun193** [comment]
Done

File: tracks/python/exercises/grains/mentoring.md (hunk 12 above)
Comment (kotp) [reply tracks/python/exercises/grains/mentoring.md:40]: This is still outstanding.

It is the code block in general, though.

Can you review this as a whole and see what you think?

```python
def square(number):
     if number < 1 or number > 64:
        raise ValueError("square must be between 1 and 64")
    else:
        return 2 << number - 1
```
The precedence is not changed by the parenthesis, and so slow down the reader to see what is special here.  It does require that one becomes familiar with order of operations for what we use, but this is not unreasonable.  It is better than overuse of parenthesis, often touted as "but for the readability" which turns out to be not so true.

File: tracks/python/exercises/grains/mentoring.md [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+```python
+def square(number):
+    if(number<1 or number>64):
+        raise ValueError("square must be between 1 and 64")
+    else:
+        return 2<<(number-1)
+
+
+def total():
+    #as 2^0+2^1+2^2+2^3+....2^n=(2^(n+1))-1
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:44]: Even as a comment, we should space our "words".

File: tracks/python/exercises/grains/mentoring.md [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    return 2<<(64)-1
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:45]:
```suggestion
    return 2 << 64 - 1
```

**Review by kotp — CHANGES_REQUESTED** [review]
For headings, leave one blank line after, and we can use code format for function names in those headers.

Sentences should start with a capital letter in English

Heading levels should be consistent, so double check.  I think the second level 2 header is probably less of a level than it should be.

File: tracks/python/exercises/grains/mentoring.md [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+# Solution using bitwise left operator.
+
+```python
+def square(number):
+    if number < 1 or number > 64:
+        raise ValueError("square must be between 1 and 64")
+    else:
+        return 1 << number - 1 
+
+
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:45]: Extra newline here.  Probably not desired.

File: tracks/python/exercises/grains/mentoring.md [hunk 17]
```diff
@@ -0,0 +1,67 @@
+# Reasonable solution by calculating power.
+
+```python
+def square(number):
+    """
+    This Function must return the number of grains on a given square number.
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:6]: "Function" is not a proper name, and should not be capitalized.  Same thing happens further on as well.

**Review by kotp — CHANGES_REQUESTED** [review]

File: tracks/python/exercises/grains/mentoring.md (hunk 16 above)
Comment (kotp) [reply tracks/python/exercises/grains/mentoring.md:45]: Also attend to trailing white space.

File: tracks/python/exercises/grains/mentoring.md [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+
+## For function `square(number)` :
+
+As we did `2**(number - 1)` on block one there will be only 1 grain, 2 to the power of 0 equals 1.
+On block 2 there will 2 and on block 3 there will be 4 which is equal to 2 to the power of 1, and 2 to the power of 2, respectively.
+Hence generalized, **2 to the power of (number-1)**.
+
+## For function `total()` :
+
+There is a general formula :
```
Comment (kotp) [inline tracks/python/exercises/grains/mentoring.md:32]: This is the first place that there is trailing white space.  Here is a diff that can be applied to address the rest of them as well:

```diff
diff --git i/tracks/python/exercises/grains/mentoring.md w/tracks/python/exercises/grains/mentoring.md
index 3e7e7ab2..5696c77a 100644
--- i/tracks/python/exercises/grains/mentoring.md
+++ w/tracks/python/exercises/grains/mentoring.md
@@ -29,7 +29,7 @@ Hence generalized, **2 to the power of (number-1)**.
 
 ## For function `total()` :
 
-There is a general formula : 
+There is a general formula :
 **2^0 + 2^1 + 2^2 + 2^3 +....+ 2^n = (2^(n + 1)) - 1**.
 
 # Solution using bitwise left operator.
@@ -39,8 +39,8 @@ def square(number):
     if number < 1 or number > 64:
         raise ValueError("square must be between 1 and 64")
     else:
-        return 1 << number - 1 
-        
+        return 1 << number - 1
+
 def total():
     # As 2^0 + 2^1 + 2^2 + 2^3 +....+ 2^n = (2^(n + 1)) - 1
     return (1 << 64) - 1
@@ -48,7 +48,7 @@ def total():
 
 # Left shift bitwise operator.
 
-**2 to the power of n = 1 << n** 
+**2 to the power of n = 1 << n**
 ```python
 >>> 2^0
 1
```

**Review by kotp — CHANGES_REQUESTED** [review]

File: tracks/python/exercises/grains/mentoring.md (hunk 18 above)
Comment (Tarun193) [reply tracks/python/exercises/grains/mentoring.md]: done

**Comment by Tarun193** [comment]
According to me it is fine now, what's your opinion.

**Review by kotp — APPROVED** [review]

File: tracks/python/exercises/grains/mentoring.md [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    if number < 1 or number > 64:
+        raise ValueError("square must be between 1 and 64")
+    else:
+        return 2**(number - 1)
+
+def total():
+    """
+    This function must return the total number of grains on the chess board.
+    :return int total number of grains a chess board can contain.
+    """
```
Comment (BethanyG) [inline tracks/python/exercises/grains/mentoring.md:19]:
```suggestion
    """Return the total number of grains on the chess board.
    
    :return int total number of grains a chess board can contain.
    """
```

File: tracks/python/exercises/grains/mentoring.md [hunk 20]
```diff
@@ -0,0 +1,62 @@
+# Reasonable solution by calculating power.
+
+```python
+def square(number):
+    """
+    This function must return the number of grains on a given square number.
+    :param: int square number on which we want to calculate the amount of grains.
+    :return int amount of grains present on the give square number(number).
+    """
```
Comment (BethanyG) [inline tracks/python/exercises/grains/mentoring.md:9]:
```suggestion
def square(number):
    """Return the number of grains on a given square number.
    
    :param: int - square number on which we want to calculate the amount of grains.
    :return int - amount of grains present on the give square number(number).
    """
```

File: tracks/python/exercises/grains/mentoring.md (hunk 19 above)
Comment (BethanyG) [inline tracks/python/exercises/grains/mentoring.md:19]:
```suggestion
    """Return the total number of grains on the chess board.
    
    :return int - total number of grains a chess board can contain.
    """
```

**Review by BethanyG — APPROVED** [review]
I think these are a solid first pass.  Nice work, @Tarun193!  🌟 
@IsaacG -- I know you mentioned in chat that you had additions you wanted to make here.  Please feel free to add them in a follow-on PR, should you see anything.

**Comment by kotp** [comment]
Thanks @Tarun193 !

**Comment by Tarun193** [comment]
Your welcome @kotp , I am very happy to contribute.

**Pull request merged**